scylla/src/transport/cluster.rs

```diff
@@ -0,0 +1,227 @@
+use crate::transport::errors::QueryError;
+use crate::transport::node::{Node, NodeConnections};
+use crate::transport::topology::{Keyspace, TopologyReader};
+use crate::transport::Compression;
+
+use futures::{future::RemoteHandle, FutureExt};
+use std::collections::{BTreeMap, HashMap};
+use std::net::SocketAddr;
+use std::sync::{Arc, RwLock};
+
+/// Cluster manages up to date information and connections to database nodes.
+/// All data can be accessed by cloning Arc<ClusterData> in the `data` field
+pub struct Cluster {
+    data: Arc<RwLock<Arc<ClusterData>>>,
+
+    refresh_channel: tokio::sync::mpsc::Sender<RefreshRequest>,
+    _worker_handle: RemoteHandle<()>,
+}
+
+pub struct ClusterData {
+    pub known_peers: HashMap<SocketAddr, Arc<Node>>, // Invariant: nonempty after Cluster::new()
+    pub ring: BTreeMap<Token, Arc<Node>>,            // Invariant: nonempty after Cluster::new()
+    pub keyspaces: HashMap<String, Keyspace>,
+}
+
+// Works in the background to keep the cluster updated
+struct ClusterWorker {
+    // Cluster data to keep updated:
+    cluster_data: Arc<RwLock<Arc<ClusterData>>>,
+
+    // Cluster connections
+    topology_reader: TopologyReader,
+    compression: Option<Compression>,
+
+    // To listen for refresh requests
+    refresh_channel: tokio::sync::mpsc::Receiver<RefreshRequest>,
+}
+
+struct RefreshRequest {
+    response_chan: tokio::sync::oneshot::Sender<Result<(), QueryError>>,
+}
+
+impl Cluster {
+    pub async fn new(
+        initial_peers: &[SocketAddr],
+        compression: Option<Compression>,
+    ) -> Result<Cluster, QueryError> {
+        let cluster_data = Arc::new(RwLock::new(Arc::new(ClusterData {
+            known_peers: HashMap::new(),
+            ring: BTreeMap::new(),
+            keyspaces: HashMap::new(),
+        })));
+
+        let (refresh_sender, refresh_receiver) = tokio::sync::mpsc::channel(32);
+
+        let worker = ClusterWorker {
+            cluster_data: cluster_data.clone(),
+
+            topology_reader: TopologyReader::new(initial_peers, compression),
+            compression,
+
+            refresh_channel: refresh_receiver,
+        };
+
+        let (fut, worker_handle) = worker.work().remote_handle();
+        tokio::spawn(fut);
+
+        let result = Cluster {
+            data: cluster_data,
+            refresh_channel: refresh_sender,
+            _worker_handle: worker_handle,
+        };
+
+        result.refresh_topology().await?;
+
+        Ok(result)
+    }
+
+    pub fn get_data(&self) -> Arc<ClusterData> {
+        self.data.read().unwrap().clone()
+    }
+
+    pub async fn refresh_topology(&self) -> Result<(), QueryError> {
+        let (response_sender, response_receiver) = tokio::sync::oneshot::channel();
+
+        self.refresh_channel
+            .send(RefreshRequest {
+                response_chan: response_sender,
+            })
+            .await
+            .unwrap_or_else(|_| panic!("Bug in Cluster::refresh_topology sending"));
```

Comment:
:+1:, It was a panic because because `SendError` didn't implement `Debug` so `expect` didn't compile, but it was because `RefreshRequest` didn't have `Debug` so I fixed that and now `expect` works.